src/core/public/application/application_service.test.ts

```diff
@@ -0,0 +1,369 @@
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { createElement } from 'react';
+import { shallow } from 'enzyme';
+
+import { injectedMetadataServiceMock } from '../injected_metadata/injected_metadata_service.mock';
+import { contextServiceMock } from '../context/context_service.mock';
+import { httpServiceMock } from '../http/http_service.mock';
+import { MockCapabilitiesService, MockHistory } from './application_service.test.mocks';
+import { MockLifecycle } from './test_types';
+import { ApplicationService } from './application_service';
+
+function mount() {}
+
+describe('#setup()', () => {
+  let setupDeps: MockLifecycle<'setup'>;
+  let startDeps: MockLifecycle<'start'>;
+  let service: ApplicationService;
+
+  beforeEach(() => {
+    setupDeps = {
+      context: contextServiceMock.createSetupContract(),
+      http: httpServiceMock.createSetupContract({ basePath: '/test' }),
+      injectedMetadata: injectedMetadataServiceMock.createSetupContract(),
+    };
+    setupDeps.injectedMetadata.getLegacyMode.mockReturnValue(false);
+    startDeps = { injectedMetadata: setupDeps.injectedMetadata };
+    service = new ApplicationService();
+  });
+
+  describe('register', () => {
+    it('throws an error if two apps with the same id are registered', () => {
+      const { register } = service.setup(setupDeps);
+
+      register(Symbol(), { id: 'app1', mount } as any);
+      expect(() =>
+        register(Symbol(), { id: 'app1', mount } as any)
+      ).toThrowErrorMatchingInlineSnapshot(
+        `"An application is already registered with the id \\"app1\\""`
+      );
+    });
+
+    it('throws error if additional apps are registered after setup', async () => {
+      const { register } = service.setup(setupDeps);
+
+      await service.start(startDeps);
+      expect(() =>
+        register(Symbol(), { id: 'app1', mount } as any)
+      ).toThrowErrorMatchingInlineSnapshot(`"Applications cannot be registered after \\"setup\\""`);
+    });
+
+    it('throws an error if an App with the same appBasePath is registered', () => {
+      const { register, registerLegacyApp } = service.setup(setupDeps);
+
+      register(Symbol(), { id: 'app1', mount } as any);
+
+      expect(() =>
+        register(Symbol(), { id: 'app2', mount, appBasePath: '/app/app1' } as any)
+      ).toThrowErrorMatchingInlineSnapshot(
+        `"An application is already registered with the appBasePath \\"/app/app1\\""`
+      );
+      expect(() => registerLegacyApp({ id: 'app1' } as any)).toThrowErrorMatchingInlineSnapshot(
+        `"An application is already registered with the appBasePath \\"/app/app1\\""`
+      );
+    });
```

Comment:
Could you add a test when apps with different id try to register the same `appBasePath ` explicitly?

Reply:
This is captured in the latest PR.

Reply:
could you link the test? I cannot find this scenario covered

Reply:
https://github.com/elastic/kibana/pull/51527/files#diff-2b0e5c210310def7eb6e39acf3d56ebfR69

Reply:
This test checks that app2 cannot register the same route that created for app1 **implicitly**. My initial comment asked to add a check for explicit routes:
```js
register(Symbol(), { id: 'app1', mount, appRoute: '/login' })
register(Symbol(), { id: 'app2', mount, appRoute: '/login' })
```